### ISSUE TITLE
Uploading data is painfully slow

### ISSUE THREAD

**Issue by ukd1** [pachyderm/pachyderm#930]
I'm attempting to upload 2.3G text file to a fresh cluster and getting ~1MB/s; is this expected?

**Comment by jdoliner** [comment]
@ukd1 how are you uploading the file? Is it with `pachctl put-file`? or through the fuse layer?
Also is `pachctl` running on the same machine as the cluster or are they separate?

Pachyderm can also pull in files via http which should be faster because the cluster machine, if it's running on some cloud provider normally has a very fast connection compared to laptops. You can do this with `pachctl put-file repo branch path -f http://example.com`?

Sorry you ran into this, we're working on hammering out our performance problems in the upcoming release.

**Comment by ukd1** [comment]
@jdoliner I'm doing it from another server (not laptop, wired in) with ``pachctl put-file`` with a 1000mbit connection on the same network....

I will try http, but I can't see that being any better connection-wise (it's 100% not this box / it's connection).

Suggestions based on that?

**Comment by msteffen** [comment]
@ukd1 thanks for filing this. I've started looking into it, and just to get a little bit more context, are you running Kubernetes on your own servers, or are you using GCP or AWS (and running the pachctl tool inside a VM)? If you're running it on your own servers, do you know offhand what you're using for block storage? Or is it something else?

Thanks!

**Comment by ukd1** [comment]
GCP, pachyderm v1.2.1, fresh install yesterday, I configured with 500g of storage; pachctl is running direct on my local server.

**Comment by ukd1** [comment]
ping? anything I can do or is this expected?

**Comment by msteffen** [comment]
@udk1 this is not expected, and I've reproduced the issue locally. We're still trying to identify the root cause (our current hypothesis is that it may something to do with the details of how we use RethinkDB since we refactored PFS in 1.2) but I'll keep updating this issue with progress as we learn more. Thanks for your patience!

**Comment by ukd1** [comment]
@msteffen ok - is there a work around? at the moment this renders pachyderm useless for me

**Comment by msteffen** [comment]
@ukd1 thanks again for being patient with us! One possible workaround that you might try is to see if you can get the file onto the same host that's running pachd by some means other than put-file (e.g. rsync) and then running `pachctl put-file` on that host with the now-local file. I haven't tried this on GCP yet, but once I do I'll post a set of commands that will do it. Obviously this isn't an ideal way to be running put-file, but our main goal is to understand the bug and get it fixed.

Just to provide some transparency into what's going on: our initial hypothesis was that pachyderm was breaking up the file too much when sending it to pachd, and that pachd was being overwhelmed by the overhead cost of processing each small chunk. After reproducing the result though, we scanned the pachyderm logs and didn't see that behavior.

Our next hypothesis was that pfs was encountering some bottleneck when storing this data to block storage. We did just rewrite PFS, so this might have explained why we didn't encounter this sooner. However, in one experiment, I created a kubernetes cluster out of a single machine and ran put-file on the same machine, and this actually ran much faster (100MB/s instead of 1MB/s). This is why I'd expect running `put-file` on the same machine as pachd to work.

My current hypothesis is that the bottleneck is related to the network connection between pachyderm and pachd, but I haven't been able to make more progress than that. I will try to keep updating this bug, though, and I'll let you know as soon as we've solved to problem, in case you want to try again.

Thanks again for reporting this, and for being patient while we work through it!

**Comment by msteffen** [comment]
@ukd1 Hi Russell, sorry for the slow update, but it turns out we were looking in completely the wrong place. It looks like at least one potential cause for this slowness in the proxy created by `pachctl port-forward`. 

Is that how you're connecting to your pachyderm cluster? If so, you may be able to fix the slowness like so:
```sh
> NODE_IP=$(kc get nodes -o='jsonpath={$.items[0].status.addresses[?(@.type=="LegacyHostIP")].address}')
> export ADDRESS=${NODE_IP}:30650
> pachctl version  # should give a version for pachd if you've connected successfully
```

This works by connecting to the first node in your kubernetes cluster listed by `kc get nodes` and then connecting to the pachd service on port `30650`. In a quick test, I got much higher throughput with `put-file` after connecting to the node directly.

If you get a chance, let me know if this seems to work for you! Our current plan is to get rid of `pachctl port-forward` in future versions, bypassing this issue.

**Comment by msteffen** [comment]
I'm closing this for now, just because I have no other remedies that I plan to try. However, @ukd1, if the `ADDRESS` fix above doesn't work for you, please re-open the issue so I can keep hunting for the problem

**Comment by derekchiang** [comment]
I think I might've stumbled upon a [gRPC issue](https://github.com/grpc/grpc-go/issues/1043) that could explain the slow uploads.

To summarize, `gRPC-go` currently uses its own incomplete implementation of HTTP/2, and it hasn't implemented dynamic flow control.  Specifically, the window size is hardcoded to 16KB (!!), which significantly limits throughput in high latency networks, since you need to pay one RTT per 16KB.

On the bright side, there seems to be [work under way](https://github.com/grpc/grpc-go/pull/1248) to address this issue, so we can probably expect a fix in the next release of gRPC.

If we don't want to wait (or if we don't trust gRPC), we can also add a new server-side implementation for `GetFile` and `PutFile` that doesn't use gRPC.  For instance, the new implementation can simply speak HTTP.  Then we can update the Go client (which is used by `pachctl`) to talk to the HTTP API.

@jdoliner @msteffen thoughts?

**Comment by derekchiang** [comment]
Since the issue is from a while ago, I think the first step is to confirm that `get-file/put-file` is still slower than what you would expect.  That seems to be the case from my personal experience, with or without port-forwarding.

**Comment by derekchiang** [comment]
So I ran some benchmarks and was able to confirm that uploading reasonably big files (10MB+) over plain HTTP is about 100% to 150% faster than uploading the same files over gRPC.

I then investigated the potential solution which is to implement `GetFile/PutFile` on HTTP.  However, there currently isn't a satisfying solution for running HTTP and gRPC servers on the same port.  See [this](https://github.com/grpc/grpc-go/issues/75) and [this](https://github.com/grpc/grpc-go/issues/549#issuecomment-191304774) and [this](https://github.com/cockroachdb/rpc-bench) for details.

Currently the best solution for multiplexing HTTP and gRPC onto the same port seems to be to use [cmux](https://github.com/cockroachdb/cmux/), which is used in etcd and CockroachDB so presumably it's a reasonably solution.  However, we consider re-implementing `GetFile/PutFile` and introducing cmux to be too high of a cost.

TL;DR: we've decided to wait for gRPC-go to fix the issue upstream, as opposed to implementing our own solution.  In the mean time, if you need the best performance for uploading files, use URLs as opposed to plain files.  That is, you'd first upload your files to the web (e.g. S3, if the file is not there already), then you'd do `put-file -f s3://...`.  That way pachd would download the file over HTTP.

**Comment by jdoliner** [comment]
At this point we've had a lot of iteration on the upload speed for Pachyderm and are pretty happy with where it's at. It can always get better but I think it makes sense to close this issue in favor of more specific performance asks.

**Comment by ajeesh-joshy** [comment]
For me even without the port forwarding, the uploading is very slow. It has 76,906 files, totalling 5.6 GB. In 1 hour it does not even complete 10% of the files. Is this expected ? I am running the cluster on the same local machine - run minikube with 8GB RAM as well. But, no notable difference.

**Comment by echohack** [comment]
> For me even without the port forwarding, the uploading is very slow. It has 76,906 files, totalling 5.6 GB. In 1 hour it does not even complete 10% of the files. Is this expected ? I am running the cluster on the same local machine - run minikube with 8GB RAM as well. But, no notable difference.

Hey @ajeesh-joshy 

This is a closed issue. Can you open a new issue for what you're encountering?

Please include the version you're running on and the details of your setup so we can help. Thank you!
